### ISSUE TITLE
Improve `uplc` parser

### ISSUE THREAD

**Issue by ChristianoBraga** [runtimeverification/plutus-core-semantics#161]
IOG's implementation of UPLC, `uplc`, does not support the concrete syntax for polymorphic types or UPLC data constructs. However, we need this support to continue to apply our test conformance strategy that essentially compares the output of `kplc` with the output of `uplc`.

**Comment by SchmErik** [comment]
I am writing down my thoughts on how this could be implemented in uplc.

What needs to be done:

1. Add support for `data` constants encoded in bytestring format
uplc's flat parser seems to have support for this format. therefore, it very likely has an internal representation.
2. Add support for `data` constants encoded in textual format
The text contained in `{}` is the result of haskell's `show` typeclass. While parsing, you may want to consider taking the contents of `{}` and doing the reverse of `show`. Note, I'm not familiar enough with haskell and the libraries used to definitively say that this would work or not.
3. Add support for parsing `list` constants and `pair` constants
The syntax of these constants were informed by writing untyped plutus programs to generate a constant from the pretty printer. It is likely that there is an internal representation of these terms but they might be buried somewhere in the interpreter code...

**Comment by ChristianoBraga** [comment]
Hey @guy.repta thanks for taking this! The assignment is to add parsing and pretty-printing support to the Untyped Plutus Core language (UPLC). Onchain contracts in the Cardano blockchain are written in this language. This is the language whose semantics is being implemented in K by the KPlutus project.

UPLC is essentially a functional language similar to untyped lambda-calculus. Similar because there is no beta-reduction, only top-level evaluation.

UPLC defines a bunch of primitives, called builtins in UPLC jargon, that handle constants. These constants are tagged as string, bool, integer, or unit. But they can also be tagged as list of something or pair of a couple of things. Moreover, they can be tagged as data, an encoded version of (a collection) of the UPLC constants.

Polymorphic datatypes and data do not have an associated concrete syntax.

@erik.kaneda has suggested one that has been accepted by the project members, both at IOG and RV, and has been implemented in KPLC.

However, we need support to this syntax also at IOG's implementation of UPLC, called uplc for our conformance test: for every test, we compare the output of uplc with the output of kplc.

Finally, the assignment is to extend uplc parser and pretty-printer with support to the syntax we defined for polymorphic types and data (assuming they have a way to handle such terms internally) so we can apply the conformance test procedure to all tests that were identified in the first task of Milestone 1.
